airbyte-queue/src/main/java/io/airbyte/queue/BigQueue.java

```diff
@@ -50,56 +68,73 @@ public BigQueue(Path persistencePath, String queueName) throws IOException {
   @Override
   public boolean offer(byte[] bytes) {
     Preconditions.checkState(!closed.get());
-    try {
-      queue.enqueue(bytes);
-    } catch (IOException e) {
-      throw new RuntimeException(e);
-    }
-    return true;
+    return BlockingOp.execute(lock, () -> {
+      try {
+        queue.enqueue(bytes);
+        return true;
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    });
   }
 
   @Override
   public byte[] poll() {
     Preconditions.checkState(!closed.get());
-    try {
-      return queue.dequeue();
-    } catch (IOException e) {
-      throw new RuntimeException(e);
-    }
+    return NonBlockingOp.execute(lock, () -> {
+      try {
+        return queue.dequeue();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    });
   }
 
   @Override
   public byte[] peek() {
     Preconditions.checkState(!closed.get());
-    try {
-      return queue.peek();
-    } catch (IOException e) {
-      throw new RuntimeException(e);
-    }
+    return NonBlockingOp.execute(lock, () -> {
+      try {
+        return queue.peek();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    });
   }
 
   @Override
   public int size() {
     Preconditions.checkState(!closed.get());
-    return Math.toIntExact(queue.size());
+    return NonBlockingOp.execute(lock, () -> Math.toIntExact(queue.size()));
   }
 
+  /**
+   * Logging frameworks call this method when printing out this class. Copy existing elements and
+   * reinsert them back into the queue so as not to empty this queue while printing it.
+   *
+   * When constructing an iterator, no other method is allowed to function to preserve order.
+   *
+   * Since this reads a disk-based queue into memory, care must be taken when retrieving this queue's
+   * iterator. Attempting to return the iterator of a queue that is too big will result in an
+   * {@link OutOfMemoryError}.
+   */
   @Override
   public Iterator<byte[]> iterator() {
     Preconditions.checkState(!closed.get());
-
-    return new AbstractIterator<>() {
-
-      @Override
-      protected byte[] computeNext() {
-        final byte[] poll = poll();
-        if (poll == null) {
-          return endOfData();
+    return BlockingOp.execute(lock, () -> {
+      try {
+        List<byte[]> elements = new ArrayList<>();
+        while (!queue.isEmpty()) {
+          elements.add(queue.dequeue());
         }
-        return poll;
+        for (byte[] e : elements) {
```

Comment:
i'm not sure that this is going to work with any sizable amount of data. once the amount of data in the queue is bigger than memory this is going to OOM. the whole point of using the disk queue though is because we expect to the queue to be bigger than memory. so if we are going to make it a constraint now that the queue has to be smaller than memory, then we've defeated the purpose of the queue. lmk if i'm misunderstanding something!

even if this doesn't oom, this does seems super expensive. 

i think we need to find a different path here. happy to brainstorm a big more about what we should do.

Reply:
It sounds like we just want to prevent unexpected data loss from calling `.iterator()` or printing it?

Can we just have those actions throw exceptions? It just takes away the footguns.

Reply:
Throwing an exception was my first instinct so happy to do that; it's a reasonable approach.

I added the locking stuff as an escape hatch in case a dev really needs to print this out.

Briefly spoke to Charles - who is in favor of switching out the inherited classes to cut this out at the source. I think that's the right approach too. What if we throw an exception for now, and add a todo to do that? The exception provides a small safeguard, which is slightly better than the huge pothole we have now.

Reply:
that works for me. doing just the exception for now to disable the footgun seems like the right level of effort.